ssm/src/test/scala/uk/gov/nationalarchives/DASSMClientTest.scala

```diff
@@ -0,0 +1,91 @@
+import org.scalatest.flatspec.AnyFlatSpec
+import io.circe.generic.auto.*
+import org.scalatest.matchers.should.Matchers.*
+import software.amazon.awssdk.services.ssm.SsmAsyncClient
+import software.amazon.awssdk.services.ssm.model.{GetParameterRequest, GetParameterResponse, Parameter, ParameterNotFoundException}
+
+import java.util.concurrent.CompletableFuture
+import scala.collection.mutable.ListBuffer
+
+class DASSMClientTest extends AnyFlatSpec:
+  case class Result(value: String)
+
+  case class TestParameter(name: String, encrypted: Boolean, value: String)
+  case class Errors(getParameter: Boolean = false)
+
+  def createClient(initial: ListBuffer[TestParameter], errors: Option[Errors] = None): SsmAsyncClient = new SsmAsyncClient {
+    override def serviceName(): String = "test"
+
+    override def close(): Unit = ()
+
+    override def getParameter(getParameterRequest: GetParameterRequest): CompletableFuture[GetParameterResponse] =
+      if errors.exists(_.getParameter) then throw new Exception("Error getting parameter") else
+      CompletableFuture.completedFuture {
+        initial.find(i => i.name == getParameterRequest.name)
+          .map { parameter =>
+            if parameter.encrypted != getParameterRequest.withDecryption() then
+              throw new Exception("Error decoding encrypted parameter")
+            GetParameterResponse.builder.parameter(Parameter.builder.value(parameter.value).build).build
+          }.getOrElse(throw ParameterNotFoundException.builder.message(s"Parameter ${getParameterRequest.name} not found").build)
+      }
+  }
+
+  "DASSMClient" should "return a decoded unencrypted parameter" in {
+    val ssmAsyncClient = createClient(ListBuffer(TestParameter("testName", false, """{"value": "testValue"}""")))
+
+    val client = DASSMClient[IO](ssmAsyncClient)
+
+    val result = client.getParameter[Result]("testName").unsafeRunSync()
+
+    result.value should equal("testValue")
+  }
+
+  "DASSMClient" should "return a decoded encrypted parameter" in {
+    val ssmAsyncClient = createClient(ListBuffer(TestParameter("testName", true, """{"value": "testValue"}""")))
+
+    val client = DASSMClient[IO](ssmAsyncClient)
+
+    val result = client.getParameter[Result]("testName", withDecryption = true).unsafeRunSync()
+
+    result.value should equal("testValue")
+  }
+
+  "DASSMClient" should "return an error if the parameter is encrypted and withDecryption is false" in {
+    val ssmAsyncClient = createClient(ListBuffer(TestParameter("testName", true, """{"value": "testValue"}""")))
+
+    val client = DASSMClient[IO](ssmAsyncClient)
+
+    val ex = intercept[Exception] {
+      client.getParameter[Result]("testName").unsafeRunSync()
+    }
+
+    ex.getMessage should equal("Error decoding encrypted parameter")
+  }
+
+  "DASSMClient" should "return an error if the parameter is not found" in {
+    val ssmAsyncClient = createClient(ListBuffer(TestParameter("testName", true, """{"value": "testValue"}""")))
```

Comment:
1. Should `"testName"` be `"invalidTestName"`? 
2. Maybe line 75 could be`client.getParameter[Result](ssmAsyncClient.name).unsafeRunSync()`?

Reply:
Similar comment to the previous one, line 70 is the valid initial state and then we're testing what happens if we pass in invalid state.